hazelcast-jet-core/src/test/java/com/hazelcast/jet/impl/connector/HazelcastConnector_RestartTest.java

```diff
@@ -0,0 +1,142 @@
+import com.hazelcast.spi.CoreService;
+import com.hazelcast.spi.MigrationAwareService;
+import com.hazelcast.spi.Operation;
+import com.hazelcast.spi.PartitionMigrationEvent;
+import com.hazelcast.spi.PartitionReplicationEvent;
+import com.hazelcast.test.HazelcastParallelClassRunner;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Future;
+
+import static com.hazelcast.jet.core.Edge.between;
+import static com.hazelcast.jet.core.TestUtil.throttle;
+import static com.hazelcast.jet.core.processor.SinkProcessors.writeListP;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.IntStream.range;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+@RunWith(HazelcastParallelClassRunner.class)
+public class HazelcastConnector_RestartTest extends JetTestSupport {
+
+    private JetInstance instance1;
+    private JetInstance instance2;
+
+    @Before
+    public void setup() {
+        JetConfig config = new JetConfig();
+        Config hazelcastConfig = config.getHazelcastConfig();
+        hazelcastConfig.addCacheConfig(new CacheSimpleConfig().setName("*"));
+        hazelcastConfig.addEventJournalConfig(new EventJournalConfig().setCacheName("stream*").setMapName("stream*"));
+        config.getHazelcastConfig().getServicesConfig().addServiceConfig(
+                new ServiceConfig()
+                        .setName("MigrationBlockingService")
+                        .setEnabled(true)
+                        .setImplementation(new MigrationBlockingService()));
+
+        instance1 = createJetMember(config);
+        instance2 = createJetMember(config);
+    }
+
+    @Test
+    public void when_iMapWrittenAndMemberShutdown_then_jobRestarts() throws Exception {
+        DAG dag = new DAG();
+        // source will emit 1 item per second
+        Vertex source = dag.newVertex("source",
+                throttle(() -> new ListSource(range(0, 1000).boxed().collect(toList())), 1));
+        Vertex sink = dag.newVertex("sink", writeListP("sink"));
+        dag.edge(between(source, sink));
+        source.localParallelism(1);
+
+        instance1.newJob(dag, new JobConfig().setAutoRestartOnMemberFailure(true));
+        // wait for the job to start producing
+        IList<Integer> sinkList = instance1.getHazelcastInstance().getList("sink");
+        assertTrueEventually(() -> assertTrue("no output to sink", sinkList.size() >= 4), 5);
+
+        // When
+        // initiate the shutdown. Thanks to the MigrationBlockingService migration will not finish
+        // before the latch is counted down.
+        Future shutdownFuture = spawn(() -> instance2.shutdown());
+
+        // Then - assert that the job stopped producing output
+        sleepSeconds(2);
+        assertFalse("node engine is running",
```

Comment:
`assertTrueEventually()` would be better here instead of a `sleep()` call.

Reply:
Doesn't work, the job continues a little while after `isRunning` becomes `true`. I think it becomes `true` quite quickly. It's here rather as a safety check, not really as a material conditional.